Radon cleanup final review

## Summary\nWrap up the radon-driven maintenance work that spans issues #1455 through #1500 by ensuring every touched module meets the target metrics and test coverage.\n\n## Acceptance criteria\n- Re-run  on each changed file and confirm no D/E blocks remain plus maintain MI B+/A when possible.\n- Run ============================= test session starts ==============================
platform darwin -- Python 3.11.12, pytest-8.4.0, pluggy-1.6.0
rootdir: /Users/munseungjin/workspace/research/qmtl
configfile: pytest.ini
plugins: anyio-4.9.0, hypothesis-6.140.2, xdist-3.8.0, asyncio-1.0.0
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
created: 10/10 workers
10 workers [1467 items]

....................sss.s............................................... [  4%]
........................................................................ [  9%]
........................................................................ [ 14%]
........................................................................ [ 19%]
........................................................................ [ 24%]
........................................................................ [ 29%]
........................................................................ [ 34%]
........................................................................ [ 39%]
........................................................................ [ 44%]
........................................................................ [ 49%]
........................................................................ [ 53%]
........................................................................ [ 58%]
........................................................................ [ 63%]
........................................................................ [ 68%]
..................s..................................................... [ 73%]
........................................................................ [ 78%]
.................F...................................................... [ 83%]
........................................................................ [ 88%]
........................................................................ [ 93%]
..............................................sss....................... [ 98%]
...........................                                              [100%]
=================================== FAILURES ===================================
_______________ test_rebalancing_shared_account_policy_violation _______________
[gw4] darwin -- Python 3.11.12 /Users/munseungjin/workspace/research/qmtl/.venv/bin/python3
  + Exception Group Traceback (most recent call last):
  |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_utils.py", line 76, in collapse_excgroups
  |     yield
  |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 177, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 772, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 344, in from_call
    |     result: TResult | None = func()
    |                              ^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 246, in <lambda>
    |     lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    |     raise exception
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    |     teardown.throw(exception)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/logging.py", line 850, in pytest_runtest_call
    |     yield
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    |     teardown.throw(exception)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/capture.py", line 900, in pytest_runtest_call
    |     return (yield)
    |             ^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    |     teardown.throw(exception)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    |     return result.get_result()
    |            ^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    |     raise exc.with_traceback(tb)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    |     res = yield
    |           ^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    |     teardown.throw(exception)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/skipping.py", line 263, in pytest_runtest_call
    |     return (yield)
    |             ^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    |     res = hook_impl.function(*args)
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/runner.py", line 178, in pytest_runtest_call
    |     item.runtest()
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pytest_asyncio/plugin.py", line 508, in runtest
    |     super().runtest()
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/python.py", line 1670, in runtest
    |     self.ihook.pytest_pyfunc_call(pyfuncitem=self)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_hooks.py", line 512, in __call__
    |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 167, in _multicall
    |     raise exception
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    |     teardown.throw(exception)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    |     return result.get_result()
    |            ^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_result.py", line 103, in get_result
    |     raise exc.with_traceback(tb)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    |     res = yield
    |           ^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pluggy/_callers.py", line 121, in _multicall
    |     res = hook_impl.function(*args)
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/_pytest/python.py", line 156, in pytest_pyfunc_call
    |     result = testfunction(**testargs)
    |              ^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/pytest_asyncio/plugin.py", line 773, in inner
    |     _loop.run_until_complete(task)
    |   File "/Users/munseungjin/.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    |     return future.result()
    |            ^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/tests/qmtl/services/gateway/test_rebalancing_route.py", line 288, in test_rebalancing_shared_account_policy_violation
    |     resp = await ctx.client.post(
    |            ^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1859, in post
    |     return await self.request(
    |            ^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1540, in request
    |     return await self.send(request, auth=auth, follow_redirects=follow_redirects)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1629, in send
    |     response = await self._send_handling_auth(
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1657, in _send_handling_auth
    |     response = await self._send_handling_redirects(
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1694, in _send_handling_redirects
    |     response = await self._send_single_request(request)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1730, in _send_single_request
    |     response = await transport.handle_async_request(request)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_transports/asgi.py", line 170, in handle_async_request
    |     await self.app(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/applications.py", line 112, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 187, in __call__
    |     raise exc
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 165, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 176, in __call__
    |     with recv_stream, send_stream, collapse_excgroups():
    |   File "/Users/munseungjin/.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/lib/python3.11/contextlib.py", line 158, in __exit__
    |     self.gen.throw(typ, value, traceback)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    |     raise exc
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 178, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/qmtl/services/gateway/api.py", line 225, in _degrade_middleware
    |     return await call_next(request)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 156, in call_next
    |     raise app_exc
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 141, in coro
    |     await self.app(scope, receive_or_disconnect, send_no_error)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    |     raise exc
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    |     await app(scope, receive, sender)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 714, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 734, in app
    |     await route.handle(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 288, in handle
    |     await self.app(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 76, in app
    |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    |     raise exc
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    |     await app(scope, receive, sender)
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 73, in app
    |     response = await f(request)
    |                ^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/fastapi/routing.py", line 301, in app
    |     raw_response = await run_endpoint_function(
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/fastapi/routing.py", line 212, in run_endpoint_function
    |     return await dependant.call(**values)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/munseungjin/workspace/research/qmtl/qmtl/services/gateway/routes/rebalancing.py", line 140, in post_rebalancing_execute
    |     status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | AttributeError: module 'starlette.status' has no attribute 'HTTP_422_UNPROCESSABLE_CONTENT'
    +------------------------------------

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/munseungjin/workspace/research/qmtl/tests/qmtl/services/gateway/test_rebalancing_route.py", line 288, in test_rebalancing_shared_account_policy_violation
    resp = await ctx.client.post(
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1859, in post
    return await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1540, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1629, in send
    response = await self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1657, in _send_handling_auth
    response = await self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1694, in _send_handling_redirects
    response = await self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_client.py", line 1730, in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/httpx/_transports/asgi.py", line 170, in handle_async_request
    await self.app(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 187, in __call__
    raise exc
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 176, in __call__
    with recv_stream, send_stream, collapse_excgroups():
  File "/Users/munseungjin/.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/lib/python3.11/contextlib.py", line 158, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    raise exc
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 178, in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/qmtl/services/gateway/api.py", line 225, in _degrade_middleware
    return await call_next(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 156, in call_next
    raise app_exc
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/base.py", line 141, in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    raise exc
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    await app(scope, receive, sender)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 714, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 734, in app
    await route.handle(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 288, in handle
    await self.app(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 76, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    raise exc
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    await app(scope, receive, sender)
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/starlette/routing.py", line 73, in app
    response = await f(request)
               ^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/fastapi/routing.py", line 301, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/.venv/lib/python3.11/site-packages/fastapi/routing.py", line 212, in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/munseungjin/workspace/research/qmtl/qmtl/services/gateway/routes/rebalancing.py", line 140, in post_rebalancing_execute
    status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'starlette.status' has no attribute 'HTTP_422_UNPROCESSABLE_CONTENT'

During handling of the above exception, another exception occurred:

gateway_app_factory = <function gateway_app_factory.<locals>.factory at 0x11d6cf600>

    @pytest.mark.asyncio
    async def test_rebalancing_shared_account_policy_violation(
        gateway_app_factory,
    ) -> None:
        writer = StubCommitLogWriter()
        policy = SharedAccountPolicyConfig(
            enabled=True,
            max_gross_notional=1000.0,
            max_net_notional=1000.0,
            min_margin_headroom=0.95,
        )
    
        async def handler(request: httpx.Request) -> httpx.Response:
            if request.method == "POST" and request.url.path == "/rebalancing/plan":
                body = _plan_response(
                    {
                        "w1": {
                            "scale_world": 1.0,
                            "scale_by_strategy": {},
                            "deltas": [
                                {"symbol": "BTCUSDT", "delta_qty": -0.4, "venue": "binance"}
                            ],
                        }
                    },
                    global_deltas=[{"symbol": "BTCUSDT", "delta_qty": -0.4, "venue": "binance"}],
                )
                return httpx.Response(200, json=body)
            if request.method == "GET" and request.url.path == "/worlds/w1":
                return httpx.Response(200, json={"world": {"id": "w1", "mode": "paper"}})
            raise AssertionError(f"Unexpected path {request.url.path}")
    
        async with gateway_app_factory(
            handler,
            commit_log_writer=writer,
            app_kwargs={"shared_account_policy_config": policy},
        ) as ctx:
>           resp = await ctx.client.post(
                "/rebalancing/execute?submit=true&shared_account=true",
                json=_default_payload(),
                headers={"X-Allow-Live": "true"},
            )

tests/qmtl/services/gateway/test_rebalancing_route.py:288: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.11/site-packages/httpx/_client.py:1859: in post
    return await self.request(
.venv/lib/python3.11/site-packages/httpx/_client.py:1540: in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/httpx/_client.py:1629: in send
    response = await self._send_handling_auth(
.venv/lib/python3.11/site-packages/httpx/_client.py:1657: in _send_handling_auth
    response = await self._send_handling_redirects(
.venv/lib/python3.11/site-packages/httpx/_client.py:1694: in _send_handling_redirects
    response = await self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/httpx/_client.py:1730: in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/httpx/_transports/asgi.py:170: in handle_async_request
    await self.app(scope, receive, send)
.venv/lib/python3.11/site-packages/fastapi/applications.py:1054: in __call__
    await super().__call__(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/applications.py:112: in __call__
    await self.middleware_stack(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/middleware/errors.py:187: in __call__
    raise exc
.venv/lib/python3.11/site-packages/starlette/middleware/errors.py:165: in __call__
    await self.app(scope, receive, _send)
.venv/lib/python3.11/site-packages/starlette/middleware/base.py:176: in __call__
    with recv_stream, send_stream, collapse_excgroups():
../../../.local/share/uv/python/cpython-3.11.12-macos-aarch64-none/lib/python3.11/contextlib.py:158: in __exit__
    self.gen.throw(typ, value, traceback)
.venv/lib/python3.11/site-packages/starlette/_utils.py:82: in collapse_excgroups
    raise exc
.venv/lib/python3.11/site-packages/starlette/middleware/base.py:178: in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
qmtl/services/gateway/api.py:225: in _degrade_middleware
    return await call_next(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/starlette/middleware/base.py:156: in call_next
    raise app_exc
.venv/lib/python3.11/site-packages/starlette/middleware/base.py:141: in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
.venv/lib/python3.11/site-packages/starlette/middleware/exceptions.py:62: in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:53: in wrapped_app
    raise exc
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:42: in wrapped_app
    await app(scope, receive, sender)
.venv/lib/python3.11/site-packages/starlette/routing.py:714: in __call__
    await self.middleware_stack(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/routing.py:734: in app
    await route.handle(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/routing.py:288: in handle
    await self.app(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/routing.py:76: in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:53: in wrapped_app
    raise exc
.venv/lib/python3.11/site-packages/starlette/_exception_handler.py:42: in wrapped_app
    await app(scope, receive, sender)
.venv/lib/python3.11/site-packages/starlette/routing.py:73: in app
    response = await f(request)
               ^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/fastapi/routing.py:301: in app
    raw_response = await run_endpoint_function(
.venv/lib/python3.11/site-packages/fastapi/routing.py:212: in run_endpoint_function
    return await dependant.call(**values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

payload = MultiWorldRebalanceRequest(total_equity=100000.0, world_alloc_before={'w1': 1.0}, world_alloc_after={'w1': 1.0}, posit...total=None, strategy_alloc_after_total=None, min_trade_notional=None, lot_size_by_symbol=None, mode=None, overlay=None)
request = <starlette.requests.Request object at 0x11d7c2210>
world_client = <qmtl.services.gateway.world_client.WorldServiceClient object at 0x11d704e50>
manager = StrategyManager(redis=<fakeredis.aioredis.FakeRedis(<redis.asyncio.connection.ConnectionPool(<fakeredis.aioredis.FakeC...ice object at 0x11d705250>, world_client=<qmtl.services.gateway.world_client.WorldServiceClient object at 0x11d704e50>)
database = StubGatewayDatabase(events=[]), enforce_live_guard = True

    @router.post("/rebalancing/execute")
    async def post_rebalancing_execute(
        payload: MultiWorldRebalanceRequest,
        request: Request,
        world_client: WorldServiceClient = Depends(deps.provide_world_client),
        manager: StrategyManager = Depends(deps.provide_manager),
        database: Database = Depends(deps.provide_database),
        enforce_live_guard: bool = Depends(deps.provide_enforce_live_guard),
    ) -> Dict[str, Any]:
        submit_requested = _parse_bool(request.query_params.get("submit"))
    
        if submit_requested and enforce_live_guard:
            if request.headers.get("X-Allow-Live") != "true":
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail={
                        "code": "E_PERMISSION_DENIED",
                        "message": "live trading not allowed",
                    },
                )
    
        # 1) Fetch plan from WorldService
        plan_resp = await world_client.post_rebalance_plan(
            payload.model_dump(exclude_unset=True)
        )
    
        # 2) Convert per-world plans into orders
        per_world_orders: Dict[str, List[dict]] = {}
        per_world_plans = plan_resp.get("per_world", {})
        venue_policies = _resolve_venue_policies(request)
        lot_sizes = payload.lot_size_by_symbol or {}
        marks_by_world, marks_global = _collect_mark_snapshots(payload.positions or [])
        current_net_notional = _aggregate_net_notional(payload.positions or [])
        min_trade_notional = payload.min_trade_notional
        for wid, p in per_world_plans.items():
            plan_obj = type(
                "_Plan",
                (),
                {
                    "world_id": wid,
                    "scale_world": p.get("scale_world", 1.0),
                    "scale_by_strategy": p.get("scale_by_strategy", {}),
                    "deltas": [
                        type(
                            "_Delta",
                            (),
                            {
                                "symbol": d.get("symbol"),
                                "delta_qty": float(d.get("delta_qty", 0.0)),
                                "venue": d.get("venue"),
                            },
                        )
                        for d in p.get("deltas", [])
                    ],
                },
            )
            orders = orders_from_world_plan(
                plan_obj,
                options=OrderOptions(
                    lot_size_by_symbol=lot_sizes,
                    min_trade_notional=min_trade_notional,
                    marks_by_symbol=marks_by_world.get(wid, {}),
                    venue_policies=venue_policies,
                ),
            )
            per_world_orders[wid] = orders
    
        # 3) Shared account global netting (optional) or overlay mode
        shared_account = request.query_params.get("shared_account", "false").lower() in {"1", "true", "yes"}
        mode = (payload.mode or 'scaling').lower() if hasattr(payload, 'mode') else 'scaling'
        orders_global: List[dict] | None = None
        policy: SharedAccountPolicy | None = getattr(
            request.app.state, "shared_account_policy", None
        )
        if shared_account:
            if policy is None or not policy.enabled:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail={
                        "code": "E_SHARED_ACCOUNT_DISABLED",
                        "message": "shared-account execution is disabled",
                    },
                )
            global_deltas = plan_resp.get("global_deltas", [])
            orders_global = orders_from_symbol_deltas([
                type("_Delta", (), {
                    "symbol": d.get("symbol"),
                    "delta_qty": float(d.get("delta_qty", 0.0)),
                    "venue": d.get("venue"),
                })
                for d in global_deltas
            ], options=OrderOptions(
                lot_size_by_symbol=lot_sizes,
                min_trade_notional=min_trade_notional,
                marks_by_symbol=marks_global,
                venue_policies=venue_policies,
            ))
            evaluation = policy.evaluate(
                orders_global,
                marks_by_symbol=marks_global,
                total_equity=payload.total_equity,
                current_net_notional=current_net_notional,
            )
            if not evaluation.allowed:
                message = evaluation.reason or "shared-account policy rejected execution"
                raise HTTPException(
>                   status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                    detail={
                        "code": "E_SHARED_ACCOUNT_POLICY",
                        "message": message,
                        "context": dict(evaluation.context),
                    },
E                   AttributeError: module 'starlette.status' has no attribute 'HTTP_422_UNPROCESSABLE_CONTENT'

qmtl/services/gateway/routes/rebalancing.py:140: AttributeError
------------------------------ Captured log call -------------------------------
WARNING  qmtl.services.gateway.api:api.py:120 Gateway events.secret not configured; using a generated secret
=========================== short test summary info ============================
FAILED tests/qmtl/services/gateway/test_rebalancing_route.py::test_rebalancing_shared_account_policy_violation
================== 1 failed, 1458 passed, 9 skipped in 15.66s ================== (or per-suite subsets if the full suite is impractical) to catch regression risks.\n- Double-check any dependency changes mentioned in the source issues and update docs/CI when needed (e.g., docs migration, mkdocs builds).\n- List the specific radon files that had hotspots and confirm in this epic’s comments when each one is verified.\n\n## Related issues\nRefs #1455 #1456 #1457 #1458 #1459 #1461 #1462 #1463 #1465 #1466 #1467 #1468 #1469 #1470 #1471 #1472 #1473 #1474 #1475 #1476 #1477 #1478 #1479 #1482 #1483 #1484 #1485 #1486 #1487 #1488 #1489 #1491 #1492 #1493 #1494 #1495 #1496 #1497 #1498 #1499 #1500

--- a/qmtl/services/gateway/routes/rebalancing.py
+++ b/qmtl/services/gateway/routes/rebalancing.py
@@ -9,6 +9,12 @@ from typing import Any, Dict, Iterable, List, Mapping, Sequence
 
 from fastapi import APIRouter, Depends, HTTPException, Request, status
 
+
+if hasattr(status, "HTTP_422_UNPROCESSABLE_CONTENT"):
+    HTTP_422_UNPROCESSABLE = status.HTTP_422_UNPROCESSABLE_CONTENT
+else:
+    HTTP_422_UNPROCESSABLE = status.HTTP_422_UNPROCESSABLE_ENTITY
+
 from .. import metrics as gw_metrics
 from ..database import Database
 from ..rebalancing_executor import (
@@ -153,7 +159,7 @@ def create_router(deps: GatewayDependencyProvider) -> APIRouter:
             if not evaluation.allowed:
                 message = evaluation.reason or "shared-account policy rejected execution"
                 raise HTTPException(
-                    status_code=status.HTTP_422_UNPROCESSABLE_CONTENT,
+                    status_code=HTTP_422_UNPROCESSABLE,
                     detail={
                         "code": "E_SHARED_ACCOUNT_POLICY",
                         "message": message,

